File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[T](list: Seq[T], func: T => Boolean): Seq[T] = list match {
    case Nil => Nil
    case head :: tail => func(head) match {
      case true => head +: keep(tail, func)
      case false => keep(tail, func)
    }
  }

  def discard[T](list: Seq[T], func: T => Boolean): Seq[T] = list match {
    case Nil => Nil
    case head :: tail => func(head) match {
      case true => discard(tail, func)
      case false => head +: discard(tail, func)
    }
  }
}

File: erlang/strain/src/strain.erl
-module(strain).

-export([keep/2, discard/2]).

keep(_, []) -> [];
keep(Fn, [Head | Tail]) ->
  case Fn(Head) of
    true -> [Head | keep(Fn, Tail)];
    false -> keep(Fn, Tail)
  end.

discard(Fn, List) ->
  keep(fun(X) -> not Fn(X) end, List).
